feat(backend/admin):  enable backups and backup management for hcp clusters (#6261)
Azure/ARO-HCP: docs/backups.md

    Admin -- "creates on-demand desires,\nreads backup status" --> Cosmos
```

### Data Flow

1. The backup schedule controller watches clusters in Cosmos DB. Once a cluster has a billing document (created after it first reaches Succeeded state) and is not marked for deletion, the controller writes Velero Schedule definitions into the kube-applier Cosmos container as ApplyDesires, and creates corresponding ReadDesires to observe their status.
2. kube-applier reads the ApplyDesires and applies the Velero Schedule resources to the appropriate management cluster in the `velero` namespace.
3. Velero executes backups according to each schedule's cadence, uploading backup data to Azure Blob Storage.
4. kube-applier reads the Velero Schedule status and writes it back into ReadDesire status in Cosmos DB.
5. The admin API reads ReadDesire statuses to serve per-schedule backup time and phase. The ServiceProviderCluster record stores only the backup schedule enabled/paused state.

## What Gets Backed Up

Each backup targets the two namespaces associated with a hosted control plane: the hosted cluster namespace and the hosted control plane namespace.

> Captured resources include:

Reviewer: I don't see related swift CRs i.e PodNetwork and PodNetworkInstance being mentioned here, is that intentional?

The PodNetwork instance is a global CR, the PodNetworkInstance lives in the HCP namespace.

Reviewer: Or the backup are not to be intended for usage of cross management cluster migration?

Author: Interesting you call that out, because I have not included it and I have not had issues recovering clusters in my personal environment so I'm assuming it is recreated by hypershift.  I will double check this.

Author: Migration is intended to be supported in the future.

Reviewer: Those CRs are created by CS (in the future it is going to be the RP); if the expectation is that the owner of those CRs recreates them in the scenario we move the HCP from one management cluster to the other then fine.

Author: That is what I would expect, from a high level I think a HCP migration would look something like:

1. Turn down API Server for HCP
2. Take backup
3. Restore on target mgmt cluster
4. External systems (RP controllers) are told the resource moved and they move CRs to the new management cluster.

> Captured resources include:

Reviewer: You probably didn't include ManifestWorks and ManagedCluster CRs (while these CRs are around) because of https://github.com/Azure/ARO-HCP/pull/6261#discussion_r3683737782 as well?

Author: I did not include CRs that were maintained externally.  Since those are maintained by ACM, Maestro, Kube Applier those systems should be responsible for recreating the CRs.

For example:

1. Maestro is down, updates aren't making it to the management cluster, Maestro MFW CRs might be deleted.
2. Hosted control plane is lost and needs recovery
3. SRE recovers from latest good backup.  Control plane returns without Maestro since we directly apply CRs.

As soon as maestro returns it would continue to manage those CRs and the cluster would be reconciled to its desired state.

Reviewer: What about a scenario we move the cluster from one management cluster to the other?

Note: This feature is unlikely to happen to happen before those components are removed but I am just floating a scenario I floated above for the PNIs CRs. Having said that, maybe for these CRs, we don't need to discuss them as they'll no longer be a concern

Author: Please see https://github.com/Azure/ARO-HCP/pull/6261#discussion_r3685898120

### Admin API

The admin API exposes HTTP endpoints for operators to inspect and control backup behavior per cluster. It reads schedule state from Cosmos DB and surfaces per-schedule status from ReadDesires.

### kube-applier

kube-applier bridges the service cluster and each management cluster. It reads ApplyDesires from Cosmos DB and applies the corresponding Velero resources on the management cluster. It also reads Velero Schedule status and writes it back to ReadDesire status, making it visible to the backup controller and admin API.

### Velero

Velero runs on each management cluster and performs the actual backup and restore operations. It is configured with the Azure plugin (Blob Storage backend) and the HyperShift plugin (HyperShift-aware backup and restore logic).

## Schedule Cadences

Two cadence tiers are available, selected at backend deployment time:

> - **Production** — Three overlapping schedules with progressively longer retention:

Reviewer 2: ```suggestion
- **production** — Three overlapping schedules with progressively longer retention:
```

  - `hourly` — cron `0 */1 * * *`, TTL 48 hours
  - `daily` — cron `0 2 * * *`, TTL 336 hours (14 days)

>   - `weekly` — cron `0 3 * * 0`, TTL 2160 hours (90 days)

Reviewer: Do we need all these schedules?

Author: Perhaps not.  I can rehash the frequency / TTLs we need.  The frequency was borrowed from ROSA HCP implementation.

Reviewer: Ah right I see; 

I bet we can re-adjust them as we go. Maybe BU will have some input on the frequency and TTL we want. 

not a blocker

> - **Testing** — A single accelerated schedule suitable for CI and development environments:

Reviewer 2: ```suggestion
- **testing** — A single accelerated schedule suitable for CI and development environments:
```

  - `5min` — cron `*/5 * * * *`, TTL 1 hour

> All schedules run with volume snapshots enabled. Retention is configured per cadence tier as above.

Reviewer 2: ```suggestion
All schedules run with volume snapshots enabled. Cadence and retention are configured via `backend.backupCadence` in [`../config/config.yaml`](../config/config.yaml) with one of the tiers above (in lowercase!).
```

## Pause and Resume

Backup schedules can be paused at two levels:

> - **Global pause** — Controlled by a backend deployment configuration value. When set, all schedules for all clusters are paused. Takes effect on the next reconciliation cycle after the backend is redeployed.

Reviewer 2: ```suggestion
- **Global pause** — Controlled by a backend deployment configuration value ( `backend.backupPaused` in [`../config/config.yaml`](../config/config.yaml) ). When set, all schedules for all clusters are paused. Takes effect on the next reconciliation cycle after the backend is redeployed.
```

> - **Global pause** — Controlled by a backend deployment configuration value ( `backend.backupPaused` in [`../config/config.yaml`](../config/config.yaml) ). When set, all schedules for all clusters are paused. Takes effect on the next reconciliation cycle after the backend is redeployed.

Reviewer: Do you see a use case where we'll want to globally pause the controller dynamically?

Author: As in some other process in the backend pauses the controller globally?  No, I don't see that use case.

It wouldn't be a bad thing to have a global kill switch via an Admin API so we can globally pause backups, it would be faster than a config change + rollout.  I am not sure where that intent would be stored though.

Reviewer: > It wouldn't be a bad thing to have a global kill switch via an Admin API so we can globally pause backups, it would be faster than a config change + rollout. I am not sure where that intent would be stored though.

This is the use case I was thinking; a global kill switch via the admin API.

Author: How?  Is there a global storage container the pause intent can be written to?

Reviewer 2: First thing: I do like the idea of an admin API kill switch, but I have to say that I don't really believe we need it (as we have the possibility of a config change + rollout). Personally, I would not add it here right now. Maybe something for the future, when we gain some production experience with this feature?

In terms of how (just for checking if my mental model is about right): couldn't the admin api write the desired state to cosmos while that backend has a controller to make sure it is applied?

> - **Per-cluster pause** — Controlled via the admin API for a specific cluster. The cluster's backup state in Cosmos DB is updated; the controller picks up the change on its next sync and updates the Velero Schedule accordingly.

Reviewer 2: ```suggestion
- **Per-cluster pause** — Controlled via the [Admin API](#admin-api-reference) for a specific cluster. The cluster's backup state in Cosmos DB is updated; the controller picks up the change on its next sync and updates the Velero Schedule accordingly.
```

```json
{"name": "<clusterID>-20260527120000", "phase": "New", "startTimestamp": "", "completionTimestamp": ""}
```

Poll status with:

```
GET .../backups/{backupName}
```

Response:
```json
{
  "resourceID": "/subscriptions/.../Microsoft.RedHatOpenShift/hcpOpenShiftClusters/mycluster",
  "backup": {
    "name": "<clusterID>-20260527120000",
    "phase": "Completed",
    "startTimestamp": "2026-05-27T12:00:05Z",
    "completionTimestamp": "2026-05-27T12:04:32Z"
  }
}
```

The `phase` field reflects the Velero Backup phase (`New`, `InProgress`, `Completed`, `Failed`, etc.). Before the backup is applied by kube-applier the phase is `New`. If the ReadDesire KubeContent cannot be unmarshaled, phase is `Unknown`.

## Infrastructure

### Storage

Backup data is stored in Azure Blob Storage. The storage account is provisioned via Bicep templates and uses Cool access tier for cost optimization and zone-redundant storage (ZRS) where available, falling back to locally-redundant storage (LRS).

### Velero Deployment

Velero is deployed to each management cluster via a Helm chart that wraps Velero's CLI-based installation in a Kubernetes Job. Two plugins are included:

> - **Azure plugin** — Provides the Azure Blob Storage backend.

Reviewer 2: Nice to have: link to those plugins. Not essential thought.
